cmd/gaia/app/export.go

```diff
@@ -128,32 +128,52 @@ func (app *GaiaApp) prepForZeroHeightGenesis(ctx sdk.Context) {
 		return false
 	})
 
-	// iterate through validators by power descending, reset bond height, update bond intra-tx counter
+	// Iterate through validators by power descending, reset bond heights, and
+	// update bond intra-tx counters.
 	store := ctx.KVStore(app.keyStake)
 	iter := sdk.KVStoreReversePrefixIterator(store, stake.ValidatorsKey)
 	counter := int16(0)
+
+	var valConsAddrs []sdk.ConsAddress
 	for ; iter.Valid(); iter.Next() {
 		addr := sdk.ValAddress(iter.Key()[1:])
 		validator, found := app.stakeKeeper.GetValidator(ctx, addr)
 		if !found {
 			panic("expected validator, not found")
 		}
+
 		validator.BondHeight = 0
 		validator.UnbondingHeight = 0
+		valConsAddrs = append(valConsAddrs, validator.ConsAddress())
+
 		app.stakeKeeper.SetValidator(ctx, validator)
 		counter++
 	}
+
 	iter.Close()
 
 	/* Handle slashing state. */
 
-	// we have to clear the slashing periods, since they reference heights
+	// remove all existing slashing periods and recreate one for each validator
 	app.slashingKeeper.DeleteValidatorSlashingPeriods(ctx)
 
+	for _, valConsAddr := range valConsAddrs {
+		sp := slashing.ValidatorSlashingPeriod{
+			ValidatorAddr: valConsAddr,
+			StartHeight:   0,
+			EndHeight:     0,
+			SlashedSoFar:  sdk.ZeroDec(),
+		}
+		app.slashingKeeper.SetValidatorSlashingPeriod(ctx, sp)
+	}
+
```

Comment:
So @cwgoes had the perspective that the correct behavior would be to have `initChain` actually set the slashing period for validator in state at the beginning of the chain.

This fix puts the responsibility on the export function to ensure that this invariant is maintained. Just checking that this is the best approach?

Reply:
`initChain` already *does* this - usually - it just doesn't do it in the case when state was exported (which is a mistake). I think either mechanism is OK, it has to be "special-cased" either way.

Reply:
Doing reseting the slashing periods via the state export is more "backwards compatible"